src/compute/src/sink/materialized_view_v2.rs

```diff
@@ -33,5 +155,875 @@ pub(super) fn persist_sink<S>(
+                desired_frontiers: OkErr::new_frontiers(),
+                persist_frontiers: OkErr::new_frontiers(),
+                batch_description: None,
+            }
+        }
+
+        fn advance_desired_ok_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            advance(&mut self.desired_frontiers.ok, frontier);
+        }
+
+        fn advance_desired_err_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            advance(&mut self.desired_frontiers.err, frontier);
+        }
+
+        fn advance_persist_ok_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.persist_frontiers.ok, frontier) {
+                self.apply_persist_frontier_advancement();
+            }
+        }
+
+        fn advance_persist_err_frontier(&mut self, frontier: Antichain<Timestamp>) {
+            if advance(&mut self.persist_frontiers.err, frontier) {
+                self.apply_persist_frontier_advancement();
+            }
+        }
+
+        /// Apply the effects of a previous `persist` frontier advancement.
+        fn apply_persist_frontier_advancement(&mut self) {
+            let frontier = self.persist_frontiers.frontier();
+
+            // We will only emit times at or after the `persist` frontier, so now is a good time to
+            // advance the times of stashed updates.
+            self.corrections.ok.advance_since(frontier.clone());
+            self.corrections.err.advance_since(frontier.clone());
+
+            // If the `persist` frontier is beyond the `lower` of the current batch description, we
+            // won't be able to append the batch, so the batch description is not valid anymore.
+            if let Some((desc, _)) = &self.batch_description {
+                if PartialOrder::less_than(&desc.lower, frontier) {
+                    self.batch_description = None;
+                }
+            }
+        }
+
+        fn absorb_batch_description(&mut self, desc: BatchDescription, cap: Capability<Timestamp>) {
+            if PartialOrder::less_than(&desc.lower, self.persist_frontiers.frontier()) {
+                return;
+            }
+
+            self.batch_description = Some((desc, cap));
+        }
+
+        async fn maybe_write_batch(&mut self) -> Option<(ProtoBatch, Capability<Timestamp>)> {
+            let (desc, _cap) = self.batch_description.as_ref()?;
+
+            // We can write a new batch if we have seen all `persist` updates before `lower` and
+            // all `desired` updates up to `upper`.
+            let persist_complete = desc.lower == *self.persist_frontiers.frontier();
+            let desired_complete =
+                PartialOrder::less_equal(&desc.upper, self.desired_frontiers.frontier());
+            if !persist_complete || !desired_complete {
+                return None;
+            }
+
+            let (desc, cap) = self.batch_description.take()?;
+
+            debug_assert_eq!(desc.lower, *self.corrections.ok.since());
```

Comment:
Probably fine perf-wise to make this a non-debug assert. If we're relying on it for correctness, probably better that it always fire before we write something bad. Not worried about this initial PR getting it wrong, more about how the code might evolve over time and someone later accidentally breaks the invariant.

Reply:
Good point. I don't think any of the `debug_assert`s here are in hot loops, so I made them all `assert`s instead.